text/0089-metrics-views-api.md

```diff
@@ -0,0 +1,264 @@
+
+For example, consider the [Prometheus histogram exposition format](https://prometheus.io/docs/concepts/metric_types/#histogram).
+The default aggregation for the Measure metric instrument is _MinMaxSumCount_.
+This aggregation computes summary statistics about (i.e. the _min_, _max_, _sum_, and _count_ of) the set of recorded values, but doesn't capture enough information to reconstruct histogram bucket counts.
+
+If instead the Measure instrument used an [_Exact_ aggregation](https://github.com/open-telemetry/opentelemetry-specification/pull/347/files#diff-5b01bbf3430dde7fc5789b5919d03001R254-R259) -- which stores the sequence of measurements without aggregating them -- the Prometheus exporter could compute the histogram for each recording interval at export time, but at the cost of significantly increased memory usage.
+
+A better solution would be to use a custom _Histogram_ aggregation, which could track bucket counts and compute summary statistics without preserving the original measurements.
+
+As the spec is written now, all metrics are aggregated following the default aggregation for the metric instrument by which they were recorded.
+By design, there is no option to change these defaults:
+
+> [Users do not have a facility in the API to select the aggregation they want for particular instruments.](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-metrics.md#aggregations)
+
+This limitation prevents users from:
+
+- Converting aggregated data to vendor-specific exposition formats
+- Configuring aggregation options, e.g. histogram bucket boundaries
+- Specifying multiple aggregations for a single metric
+
+In addition, it's often desirable to aggregate metrics with respect to a specific set of label keys to reduce memory use and export data size.
+
+A _view API_ would allow users to use custom aggregation types, associate individual metrics with one or more aggregations, configure aggregation options, and specify the set of label keys to use to aggregate each metric.
+
+## Explanation
+
+Metric instruments are API objects that record measurements from application code.
+These measurements are aggregated by _Aggregators_, which are SDK objects.
+Aggregated measurement data are passed to _Exporters_, which convert these data into APM-specific exposition formats.
+
+Each view describes a relationship between a metric instrument and an aggregation, and specifies the set of label keys to preserve in the aggregated data.
+
+Views are tracked in a registry, which exporters may use to get the current aggregate values for all registered aggregations at export time.
+Each registered view represents a metric that is meant to be exported.
+As a corollary, the SDK can choose to drop measurements from any metric instrument that does not appear in a registered view.
```

Comment:
should behavior for the opentelemetry-sdk be stated here? Seems like we should make a choice one way or the other.

I would opt to drop measurements, as it could be significant memory overhead to just keep measurements that will never be exported.

Reply:
+1 to not having the SDK be able to "choose" but that being the required way it works.

This does not break with the desire to not require users to define views to get metrics if we go with my suggestion of default view generation for each instrument -- since there would by default be a view, so measurements would not be dropped without the end user/operator explicitly disable default views or setting their own.

Reply:
I'd like to get more clarity on "Should exporters be even aware of ViewRegistry?" Or only Meter implementation need to be aware of ViewRegistry.

Exporters will continue to simply have a method like Export(collection<metrics>). And Meter implementation takes of dropping/retaining dimensions, choosing right aggregation etc from ViewRegistry.

Reply:
I believe from last week's metrics SIG, it was decided that any measurement that was not registered with a View would be aggregated with a default aggregation.

Reply:
Curious how that works in practice. It would require *something* to iterate through all metrics during every export and see if they have a view already registered.